### PR TITLE
Replace Newtonsoft.Json with System.Text.Json in PSConfiguration

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Use System.Text.Json instead of Newtonsoft.Json in PSConfiguration.

Perf win is ~11% in startup scenario (pwsh -c exit) with two config files (user and system). It is ~60 ms on my system.

Breaking change because System.Text.Json have many minor (by design) breaking changes compared with Netonsoft.Json.
So I had to change `"true"` on `true` in powershell.config.json file in tests.

Before the PR:
image

After the PR:
image

## PR Context

Related #14268

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/PSConfiguration.cs [hunk 1]
```diff
@@ -473,83 +492,62 @@ private void UpdateValueInFile<T>(ConfigScope scope, string key, T value, bool a
...
-                            if (addValue)
-                            {
-                                propertyToModify.Replace(new JProperty(key, value));
-                            }
-                            else
-                            {
-                                propertyToModify.Remove();
-                            }
-                        }
+                var v = JsonSerializer.Serialize(value);
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/PSConfiguration.cs:511]: These conversions between `JsonElement` and strongly-typed objects here and in `ToObject<T>()` feel wrong — there must be a better way to do this

File: src/System.Management.Automation/engine/PSConfiguration.cs [hunk 2]
```diff
@@ -473,83 +492,62 @@ private void UpdateValueInFile<T>(ConfigScope scope, string key, T value, bool a
...
+                }
+                else
+                {
+                    // The property doesn't exist, so add it
+                    if (addValue)
+                    {
+                        configScopeData.Add(key, jsonElementValue);
                     }
                 }
+                var options = new JsonSerializerOptions
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/PSConfiguration.cs:535]:
```suggestion

                var options = new JsonSerializerOptions
```

**Comment by iSazonov** [comment]
Does Team want this in 7.2? 
We could benefit from JsonNode (new in .Net 6.0).

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 3]
```diff
@@ -43,6 +43,10 @@ internal static void Init()
             //   * have high disk cost
 
             // We shouldn't create too many tasks.
+            Task.Run(() =>
+            {
+                var jsonDocument = System.Text.Json.Nodes.JsonObject.Parse("{\"test\":  \"test\"}");
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/InitialSessionState.cs:48]: Is the purpose of this task just to load the assemblies?  If so, can you add a comment and you can just use the string `"1"` as valid json

File: src/System.Management.Automation/engine/PSConfiguration.cs [hunk 4]
```diff
@@ -473,83 +496,41 @@ private void UpdateValueInFile<T>(ConfigScope scope, string key, T value, bool a
...
+                if (!string.IsNullOrEmpty(jsonString))
                 {
-                    // Safely determines whether there is content to read from the file
-                    bool isReadSuccess = jsonReader.Read();
-                    if (isReadSuccess)
-                    {
-                        // Read the stream into a root JObject for manipulation
-                        jsonObject = serializer.Deserialize<JObject>(jsonReader);
-                        JProperty propertyToModify = jsonObject.Property(key);
+                    configScopeData = configScopeData = JsonNode.Parse(jsonString) as JsonObject;
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/PSConfiguration.cs:507]: Why is `configScopeData` here twice?

File: test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1 [hunk 5]
```diff
@@ -492,23 +492,23 @@ Describe "Additional tests for Import-Module with WinCompat" -Tag "Feature" {
         It "Fails to import incompatible module if implicit WinCompat is disabled in config" {
             $LogPath = Join-Path $TestDrive (New-Guid).ToString()
             $ConfigPath = Join-Path $TestDrive 'powershell.config.json'
-            '{"DisableImplicitWinCompat" : "True"}' | Out-File -Force $ConfigPath
+            '{"DisableImplicitWinCompat" : true}' | Out-File -Force $ConfigPath
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Core/CompatiblePSEditions.Module.Tests.ps1:495]: This looks like a mistake previously where it should have always been `true` and not the string `"True"`.  I think this is the only boolean setting in the config file and is not set by default, so it's unlikely to impact users.

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/InitialSessionState.cs:48]: This is for early loading.
Comment is above.
"test" was replaced with "1".

File: src/System.Management.Automation/engine/PSConfiguration.cs (hunk 4 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/PSConfiguration.cs:504]: Typo. Fixed.

File: src/System.Management.Automation/engine/PSConfiguration.cs [hunk 6]
```diff
@@ -1,16 +1,18 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-using System;
+#nullable enable
+
 using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
 using System.IO;
+#if UNIX
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/PSConfiguration.cs:9]: I don't see below why this is only for UNIX?

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/PSConfiguration.cs (hunk 6 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/PSConfiguration.cs]: It is from old code. Removed.

**Comment by iSazonov** [comment]
Rebased to move to .Net 7 RC1.

File: src/System.Management.Automation/engine/PSConfiguration.cs (hunk 6 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/PSConfiguration.cs:9]: Ah, it is still needed - line 249 (Unix logging).

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

**Review by daxian-dbw — CHANGES_REQUESTED** [review]
Not finished reviewing but want to share my comments so far.

**Comment by daxian-dbw** [comment]
I don't fully understand the gains of do this work, especially given that it's a breaking change.

> Breaking change because System.Text.Json have many minor (by design) breaking changes compared with Netonsoft.Json.
So I had to change "true" on true in powershell.config.json file in tests.

- From the perspective of startup time improvement, how much of the gain is related to spinning up a task to warm up JSON parsing in `InitialSessionState.cs`? Also, it's hard to reason from the attached screenshots in the PR description. Some benchmarking similar to https://github.com/PowerShell/PowerShell/pull/8341#issue-384623150 would be helpful to see the perf difference.

- From the perspective of mitigating assembly loading conflicts, we still have a hard dependency to `Newtonsoft.Json` in the Utility module, so switching to `System.Text.Json` in PSConfiguraiton handling doesn't really solve the problem.

**Comment by iSazonov** [comment]
> I don't fully understand the gains of this work, especially given that it's a breaking change.

Intention is to remove dependency on Newtonsoft.Json at all. It is impossible to do this in one PR, only step by step. 
Today we have active PR for Test-Json. Then we could update other two cmdlets.

Before I continue with the PR. I'd want to get a conclusion from MSFT team - do we want to rid of Newtonsoft.Json in the milestone?

**Comment by daxian-dbw** [comment]
> do we want to rid of Newtonsoft.Json in the milestone?

I think this largely depends on whether we can retain the same behavior as of today. If breaking changes are inevitable, then I personally tend to keeping it as is, but of course, it would be helpful to understand the scale and risky level of those breaking changes.

**Comment by iSazonov** [comment]
> I think this largely depends on whether we can retain the same behavior as of today.

.Net implementation fully follows JSON standard. So users cannot say that PowerShell cmdlets do not work correctly.
We have a lot of tests to help us understand how many differences there are. See #18141 for example.

Current JSON cmdlets could be published on PowerShellGet site in a compatibility module.

> Some modules assume that PowerShell 7.x ships with Newtonsoft.Json, and removing that dependency is a breaking change to those modules.

It seems .Net continue distribute Newtonsoft.Json.

**Comment by floh96** [comment]
@daxian-dbw can you mark this issue for Committee Review to get a conclusion and move this forward?

**Comment by daxian-dbw** [comment]
@PowerShell/powershell-committee Please review to decide whether this PR is desired. Please see the comments starting from https://github.com/PowerShell/PowerShell/pull/14411#issuecomment-1258755659 for reference and context.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this.  We agree that this is the right change towards removing dependency on NewtonSoft the for the engine to read `powershell.config.json`.  The breaking change for `powershell.config.json` seems unlikely to hit real world scenarios due to differences we understand for NewtonSoft.  Additional changes in JSON cmdlets, for example, need to be evaluated on a case-by-case basis so this statement isn't an automatic acceptance of further changes.  We also expect that we cannot immediately remove shipping NewtonSoft even if PS7 itself removes all dependencies and need a migration story for modules (and potentially scripts) that may explicitly depend on it.

**Comment by iSazonov** [comment]
After merging SG for PSVersionTable the config reading takes 20% in startup scenario.

**Pull request closed without merging**